index.html

```diff
@@ -872,6 +1318,115 @@ <h4>
         </section>
       </section>
     </section>
+    <section data-cite="file-system-access permissions">
+      <h2>
+        Launch queue and launch params
+      </h2>
+      <section data-dfn-for="LaunchParams">
+        <h3>
+          <dfn>LaunchParams</dfn> interface
+        </h3>
+        <pre class="idl">
+          [Exposed=Window] interface LaunchParams {
+            readonly attribute DOMString? targetURL;
+            readonly attribute FrozenArray&lt;FileSystemHandle&gt; files;
+          };
+        </pre>
+        <p>
+          {{LaunchParams/targetURL}} represents the [=URL=] of the launch which
+          MUST be [=manifest/within scope=] of the application.
+        </p>
+        <p>
+          For every |file handle:FileSystemHandle| in {{LaunchParams/files}},
+          querying the file system permission with
+          {{FileSystemPermissionDescriptor/mode}} set to
+          {{FileSystemPermissionMode/"readwrite"}} MUST return
+          {{PermissionState/"granted"}}.
+        </p>
+      </section>
+      <section data-dfn-for="LaunchConsumer">
+        <h3>
+          <dfn>LaunchConsumer</dfn> function
+        </h3>
+        <pre class="idl">
+          callback LaunchConsumer = any (LaunchParams params);
+        </pre>
+      </section>
+      <section data-dfn-for="LaunchQueue">
+        <h3>
+          <dfn>LaunchQueue</dfn> interface
+        </h3>
+        <pre class="idl">
+          partial interface Window {
+            readonly attribute LaunchQueue launchQueue;
+          };
+
+          [Exposed=Window] interface LaunchQueue {
+            undefined setConsumer(LaunchConsumer consumer);
+          };
+        </pre>
+        <p>
+          {{LaunchQueue}} has an <dfn>unconsumed launch params</dfn> which is a
+          [=list=] of {{LaunchParams}} that is initially empty.
+        </p>
+        <p>
+          {{LaunchQueue}} has an <dfn>assigned launch consumer</dfn> which is
+          initially null.
+        </p>
+        <section>
+          <h2>
+            <dfn>setConsumer</dfn> method
+          </h2>
+          <p>
+            The {{LaunchQueue/setConsumer(consumer)}} method steps are:
+          </p>
+          <ol class="algorithm">
+            <li>Set the [=assigned launch consumer=] to |consumer|.
+            </li>
+            <li>[=list/For each=] |launch_params:LaunchParams| of [=unconsumed
+            launch params=]:
+              <ol>
+                <li>Invoke |consumer| with |launch_params|.
+                </li>
+              </ol>
+            </li>
+            <li>Set [=unconsumed launch params=] to the empty [=list=].
+            </li>
+          </ol>
+        </section>
+      </section>
+      <section>
+        <h3>
+          Handling Web App Launches
```

Comment:
I wonder if we need to acknowledge the step of 'creating a new document' or browsing context.

"user agent SHOULD navigate to url and the appropriate browsing context is set to a new top level browsing context."  (same line as in protocol handling) somewhere in the algorithm?

Should we be mentioning edge cases like redirection? Or are we not mentioning that on purpose?

Reply:
I've added a line about the creation of the new browser context. I didn't use the exact same language since it's a step in an algorithm rather than part of a paragraph.

Redirection should be mentioned but I think that's sort of out of the scope of file handling. It should definitely be mentioned when the LaunchQueue/Params spec is fleshed out.